src/main/java/frc/robot/subsystems/vision/Limelight.java

```diff
@@ -0,0 +1,120 @@
+package frc.robot.subsystems.vision;
+
+import edu.wpi.first.math.geometry.Pose2d;
+import edu.wpi.first.math.geometry.Pose3d;
+import edu.wpi.first.math.geometry.Rotation2d;
+import edu.wpi.first.math.geometry.Translation2d;
+import edu.wpi.first.net.PortForwarder;
+import edu.wpi.first.networktables.*;
+import frc.robot.subsystems.LoggedSubsystem;
+
+import java.util.Optional;
+import java.util.OptionalDouble;
+
+public class Limelight extends LoggedSubsystem<LimelightLogInputs> {
+    public static Limelight INSTANCE = null;
+
+    private final NetworkTable table = NetworkTableInstance.getDefault().getTable("limelight");
+    private final DoubleSubscriber tx = table.getDoubleTopic("tx").subscribe(0.0);
+    private final DoubleSubscriber ty = table.getDoubleTopic("ty").subscribe(0.0);
+    private final IntegerSubscriber tv = table.getIntegerTopic("tv").subscribe(0);
+    private final DoubleSubscriber ts = table.getDoubleTopic("ts").subscribe(0.0);
+    private final IntegerSubscriber tid = table.getIntegerTopic("tid").subscribe(0);
+
+    private Limelight() {
+        super(new LimelightLogInputs());
+        for (int i=5800; i<=5805; i++){
+            PortForwarder.add(i, "limelight.local", i);
+        }
+    }
+
+    /**
+     * If there is no instance of the limelight class it creates one and returns it
+     * @return Limelight instance
+     */
+    public static Limelight getInstance() {
+        if (INSTANCE == null) {
+            INSTANCE = new Limelight();
+        }
+        return INSTANCE;
+    }
+
+    /**
+     * Checks weather the limelight can see any targets
+     * @return weather the limelight detects any targets
+     */
+    public boolean hasTargets() {
+        return tv.get() == 1;
+    }
+
+    /**
+     * @return aprilTag id
+     */
+    public long getTagId() {
+        return tid.get();
+    }
+
+
+    /**
+     * @return pitch from robot to target
+     */
+    public Rotation2d getPitch() {
+        return Rotation2d.fromDegrees(ty.get());
+    }
+
+    /**
+     * @return distance from target
+     */
+    public OptionalDouble getTargetDistance(double targetHeight) {
+        double totalPitch = Constants.CAMERA_PITCH + getPitch().getRadians();
+        double error = Constants.CAMERA_HEIGHT-targetHeight;
+        if (hasTargets()) {
+            return OptionalDouble.of(Math.abs(error / Math.tan(totalPitch)));
```

Comment:
Maybe I don't understand something, but shouldn't sin be used instead of tan?

Reply:
No its supposed to be tan but the calculation is wrong because I used ty (from the network tables).
The calculation should be the error * tan(camera pitch)

Reply:
I think you have a mistake, according to the [documentation](https://docs.limelightvision.io/en/latest/cs_estimating_distance.html#using-a-fixed-angle-camera) of limelight, the code is correct